File: src/org/jgll/grammar/slot/NonterminalGrammarSlot.java
package org.jgll.grammar.slot;

import java.io.IOException;
import java.io.Writer;
import java.util.BitSet;
import java.util.Set;

import org.jgll.grammar.HeadGrammarSlot;
import org.jgll.grammar.SlotAction;
import org.jgll.grammar.Symbol;
import org.jgll.grammar.Terminal;
import org.jgll.parser.GLLParser;
import org.jgll.recognizer.GLLRecognizer;
import org.jgll.util.Input;


/**
 * A grammar slot immediately before a nonterminal.
 *
 * @author Ali Afroozeh
 *
 */
public class NonterminalGrammarSlot extends BodyGrammarSlot {
	
	private static final long serialVersionUID = 1L;

	private HeadGrammarSlot nonterminal;
	
	private BitSet testSet;
	
	private int minInputVal = Integer.MAX_VALUE;
	
	private int maxInputVal;
	
	public NonterminalGrammarSlot(String label, int position, BodyGrammarSlot previous, HeadGrammarSlot nonterminal, HeadGrammarSlot head) {
		super(label, position, previous, head);
		if(nonterminal == null) {
			throw new IllegalArgumentException("Nonterminal cannot be null.");
		}
		this.nonterminal = nonterminal;
		testSet = new BitSet();
	}
	
	public HeadGrammarSlot getNonterminal() {
		return nonterminal;
	}
	
	public void setNonterminal(HeadGrammarSlot nonterminal) {
		this.nonterminal = nonterminal;
	}
	
	@Override
	public GrammarSlot parse(GLLParser parser, Input input) {
		
		for(SlotAction<Boolean> preCondition : preConditions) {
			if(!preCondition.execute(parser, input)) {
				return null;
			}
		}
		
		int ci = parser.getCurrentInputIndex();
		if(checkAgainstTestSet(ci, input)) {
			parser.createGSSNode(next);
			return nonterminal;
		} else {
			parser.recordParseError(this);
			return null;
		}		
	}
	
	@Override
	public GrammarSlot recognize(GLLRecognizer recognizer, Input input) {
		int ci = recognizer.getCi();
		org.jgll.recognizer.GSSNode cu = recognizer.getCu();
		if(checkAgainstTestSet(ci, input)) {
			recognizer.update(recognizer.create(next, cu, ci), ci);
			return nonterminal;
		} else {
			recognizer.recognitionError(cu, ci);
			return null;
		}		

	}
	
	@Override
	public void codeParser(Writer writer) throws IOException {
		
		if(previous == null) {
			codeIfTestSetCheck(writer);
			writer.append("   cu = create(grammar.getGrammarSlot(" + next.id + "), cu, ci, cn);\n");
			writer.append("   label = " + nonterminal.getId() + ";\n");
			codeElseTestSetCheck(writer);
			writer.append("}\n");
			
			writer.append("// " + next + "\n");
			writer.append("private void parse_" + next.id + "() {\n");
			
			BodyGrammarSlot slot = next;
			while(slot != null) {
				slot.codeParser(writer);
				slot = slot.next;
			}
		} 
		
		else { 
		
			// TODO: add the testSet check
			// code(A ::= α · Xl β) = 
			//						if(test(I[cI ], A, Xβ) {
			// 							cU :=create(RXl,cU,cI,cN); 
			//							gotoLX 
			//						}
			// 						else goto L0
			// RXl:
			codeIfTestSetCheck(writer);
			writer.append("   cu = create(grammar.getGrammarSlot(" + next.id + "), cu, ci, cn);\n");
			writer.append("   label = " + nonterminal.getId() + ";\n");
			codeElseTestSetCheck(writer);
			writer.append("}\n");
			
			writer.append("// " + next + "\n");
			writer.append("private void parse_" + next.id + "(){\n");
		}
	}
	
	@Override
	public void codeIfTestSetCheck(Writer writer) throws IOException {
		writer.append("if (");
//		int i = 0;
//		for(Terminal terminal : testSet) {
//			writer.append(terminal.getMatchCode());
//			if(++i < testSet.size()) {
//				writer.append(" || ");
//			}
//		}
		writer.append(") {\n");
	}

	@Override
	public boolean checkAgainstTestSet(int index, Input input) {
		int i = input.charAt(index);
		if(i < minInputVal || i > maxInputVal) {
			return false;
		}
		return testSet.get(i);
	}

	public void setTestSet(Set<Terminal> testSet) {
		if(testSet == null || testSet.isEmpty()) {
			throw new IllegalArgumentException("Test set for the nontermianl " + nonterminal.getNonterminal().getName() + " is empty");
		}
		
		for(Terminal t : testSet) {
			if(minInputVal > t.getMinimumValue()) {
				minInputVal = t.getMinimumValue();
			}
			if(maxInputVal < t.getMaximumValue()) {
				maxInputVal = t.getMaximumValue();
			}
			this.testSet.or(t.getTestSet());
		}
	}
	
	@Override
	public boolean isNullable() {
		return nonterminal.isNullable();
	}

	@Override
	public Symbol getSymbol() {
		return nonterminal.getNonterminal();
	}

	
}

File: src-scala/org/jgll_staged/grammar/slot/NonterminalGrammarSlot.scala
package org.jgll_staged.grammar.slot

import java.io.IOException
import java.io.Writer
import java.util.BitSet
import org.jgll_staged.grammar.Symbol
import org.jgll_staged.grammar.Terminal
import org.jgll_staged.parser.GLLParserInternals
import org.jgll_staged.recognizer.GLLRecognizer
import org.jgll_staged.util.Input
//remove if not needed
import scala.collection.JavaConversions._

@SerialVersionUID(1L)
class NonterminalGrammarSlot(position: Int, 
    previous: BodyGrammarSlot, 
    protected var nonterminal: HeadGrammarSlot, 
    head: HeadGrammarSlot) extends BodyGrammarSlot(position, previous, head) {

  private var firstSet: BitSet = new BitSet()

  private var followSet: BitSet = new BitSet()

  if (nonterminal == null) {
    throw new IllegalArgumentException("Nonterminal cannot be null.")
  }

  def copy(previous: BodyGrammarSlot, nonterminal: HeadGrammarSlot, head: HeadGrammarSlot): NonterminalGrammarSlot = {
    val slot = new NonterminalGrammarSlot(this.position, previous, nonterminal, head)
    slot.preConditions = preConditions
    slot.popActions = popActions
    slot.firstSet = firstSet
    slot.followSet = followSet
    slot
  }

  def getNonterminal(): HeadGrammarSlot = nonterminal

  def setNonterminal(nonterminal: HeadGrammarSlot) {
    this.nonterminal = nonterminal
  }

  override def parse(parser: GLLParserInternals, input: Input): GrammarSlot = {
    val ci = parser.getCurrentInputIndex
    if (!testFirstSet(ci, input) && !(isNullable && testFollowSet(ci, input))) {
      parser.recordParseError(this)
      return null
    }
    if (executePreConditions(parser, input)) {
      return null
    }
    parser.createGSSNode(next)
    nonterminal
  }

  override def recognize(recognizer: GLLRecognizer, input: Input): GrammarSlot = {
    val ci = recognizer.getCi
    val cu = recognizer.getCu
    if (testFirstSet(ci, input)) {
      recognizer.update(recognizer.create(next, cu, ci), ci)
      nonterminal
    } else if (isNullable && testFollowSet(ci, input)) {
      next
    } else {
      recognizer.recognitionError(cu, ci)
      null
    }
  }

  override def codeParser(writer: Writer) {
    if (previous == null) {
      codeIfTestSetCheck(writer)
      writer.append("   cu = create(grammar.getGrammarSlot(" + next.id + "), cu, ci, cn);\n")
      writer.append("   label = " + nonterminal.getId + ";\n")
      codeElseTestSetCheck(writer)
      writer.append("}\n")
      writer.append("// " + next + "\n")
      writer.append("private void parse_" + next.id + "() {\n")
      var slot = next
      while (slot != null) {
        slot.codeParser(writer)
        slot = slot.next
      }
    } else {
      codeIfTestSetCheck(writer)
      writer.append("   cu = create(grammar.getGrammarSlot(" + next.id + "), cu, ci, cn);\n")
      writer.append("   label = " + nonterminal.getId + ";\n")
      codeElseTestSetCheck(writer)
      writer.append("}\n")
      writer.append("// " + next + "\n")
      writer.append("private void parse_" + next.id + "(){\n")
    }
  }

  override def codeIfTestSetCheck(writer: Writer) {
    writer.append("if (")
    writer.append(") {\n")
  }

  def setTestSet() {
    for (t <- nonterminal.getFirstSet) {
      firstSet.or(t.asBitSet())
    }
    for (t <- nonterminal.getFollowSet) {
      followSet.or(t.asBitSet())
    }
  }

  override def testFirstSet(index: Int, input: Input): Boolean = firstSet.get(input.charAt(index))

  override def testFollowSet(index: Int, input: Input): Boolean = followSet.get(input.charAt(index))

  override def isNullable(): Boolean = nonterminal.isNullable

  override def getSymbol(): Symbol = nonterminal.getNonterminal

  override def isNameEqual(slot: BodyGrammarSlot): Boolean = {
    if (this == slot) {
      return true
    }
    if (!(slot.isInstanceOf[NonterminalGrammarSlot])) {
      return false
    }
    val other = slot.asInstanceOf[NonterminalGrammarSlot]
    nonterminal.getNonterminal == other.nonterminal.getNonterminal
  }
}
